File: ts/implementation/faster_download.ts
/**
 * function to determine the chunk size to download.
 * 
 * @param size The size of the file.
 */
export type ChunkSizeDeterminer = (size: number) => number;

/**
 * Function to determine the number of parts to download.
 * 
 * @param size The size of the file.
 */
export type PartDeterminer = (size: number) => number;

/**
 * Config for the downloader.
 */
export type DownloaderConfig = {
    /**
     * The maximum number of retries for a chunk.
     */
    maxRetries?: number;

    /**
     * The maximum number of chunks to download at once.
     */
    maxChunkSize?: number;

    /**
     * The number of chunks to download at once.
     */
    concurrency: number;

    /**
     * The function to determine the number of parts to download.
     */
    chunkSizeDeterminer?: ChunkSizeDeterminer;

    /**
     * The function to determine the chunk size to download.
     */
    partDeterminer?: PartDeterminer;

    /**
     * The actual chunk size to download.
     */
    chunkSize: number;
};


export interface WriteAt {
    WriteAt(bytes: Uint8Array, offset: number): Promise<number>;
}

export enum SizeUnit {
    B = 1,
    KB = 1024,
    MB = 1024 * 1024,
    GB = 1024 * 1024 * 1024
}

const defaultPartDeterminer: PartDeterminer = (size: number) => {
    if (size < SizeUnit.MB) {
        return 1;
    }

    if (size < 10 * SizeUnit.GB) {
        return 4;
    }

    if (size < 100 * SizeUnit.GB) {
        return 16;
    }

    return 32;
}

const defaultConfig: DownloaderConfig = {
    maxRetries: 5,
    maxChunkSize: SizeUnit.MB,
    concurrency: 4,
    partDeterminer: defaultPartDeterminer,
    chunkSize: 0
}

export class downloadChunk {
    start: number;
    end: number;
    size: number;
    cur: number;
    writer: WriteAt;

    constructor(start: number, end: number, size: number) {
        this.start = start;
        this.end = end;
        this.size = size;
        this.cur = start;
    }

    async write(bytes: Uint8Array): Promise<number> {
        if (this.cur + bytes.length > this.end) {
            throw new Error('Invalid write');
        }

        const byteWrote =  await this.writer.WriteAt(bytes, this.start+this.cur);
        this.cur += byteWrote;
        return byteWrote;
    }

    byteRage(): string {
        return `bytes=${this.start}-${this.end}`;
    }
}

class downloader {
    constructor(private config: DownloaderConfig) { }
}

export class Downloader {
    private downloader: downloader;
    constructor(private config: DownloaderConfig) {
        this.config = Object.assign({}, defaultConfig, config);
        this.downloader = new downloader(this.config);
    }
}

File: go/implementation/faster_download.go
package implementation

import "context"

// ChunkSizeDeterminer is a function that determines the chunk size based on the file size
type ChunkSizeDeterminer func(fileSize int64) int64

// PartDeterminer is a function that determines the number of parts to download concurrently
type PartDeterminer func(fileSize int64) int64

// DownloaderConfig is a struct that contains configuration for the Downloader
type DownloaderConfig struct {
	// MaxRetries is the maximum number of retries for a file to download
	// If MaxRetries is -1, it will retry forever
	//
	// If one of chunk download fails, it will retry entire file download
	MaxRetries int

	// MaxChunkSize is the maximum size of a chunk in bytes
	// If MaxChunkSize is -1, it will download the entire file in one chunk
	//
	// Default is 1MB
	MaxChunkSize int

	// The Concurrency is the number of chunks to download concurrently
	//
	// Default: -1 (download all chunks concurrently)
	Concurrency int

	// ChunkSizeDeterminer is a function that determines the chunk size based on the file size
	// If ChunkSizeDeterminer is nil, it will use the default chunk size
	//
	// Default: 1MB
	// The ChunkSize after calling ChunkSizeDeterminer will be capped by MaxChunkSize -> min(ChunkSize, MaxChunkSize)
	//
	// You should only use ChunkSizeDeterminer either PartDeterminer, not both
	ChunkSizeDeterminer ChunkSizeDeterminer

	// PartDeterminer is a function that determines the number of parts to download concurrently
	// If PartDeterminer is nil, it will use the default number of parts -> ceil(fileSize / chunkSize)
	//
	// If the number of parts is greater than Concurrency, some parts will be downloaded concurrently, and some parts will be downloaded sequentially
	//
	// Example:
	// Concurrency = 3
	// PartDeterminer = func(fileSize int64) int64 {
	// 	return 2
	// }
	// The first 2 parts will be downloaded concurrently, and the last part will be downloaded sequentially
	//
	// You should only use PartDeterminer either ChunkSizeDeterminer, not both
	//
	// By default,  Downloader will use the default PartDeterminer
	PartDeterminer PartDeterminer
}

// File Size Unit Enum
const (
	_ = 1 << (10 * iota)
	KB
	MB
	GB
	TB
	PB
	EB
)

var defaultPartDeterminer = func(fileSize int64) int64 {
	if fileSize < 1*MB {
		return 1
	}

	if fileSize < 10*MB {
		return 2
	}

	if fileSize < 100*MB {
		return 16
	}

	return 32
}

var defaultConfig = DownloaderConfig{
	MaxRetries:          5,
	MaxChunkSize:        1 * MB,
	Concurrency:         -1,
	ChunkSizeDeterminer: nil,
	PartDeterminer:      defaultPartDeterminer,
}

type Downloader struct {
	config DownloaderConfig
}

func NewDownloader(options ...Option) *Downloader {
	downloader := Downloader{config: defaultConfig}
	for _, option := range options {
		option(&downloader)
	}
	return &downloader
}

func NewDownloaderWithConfig(config DownloaderConfig) *Downloader {
	return &Downloader{config: config}
}

type Option func(d *Downloader)

func WithMaxRetries(maxRetries int) Option {
	return func(d *Downloader) {
		d.config.MaxRetries = maxRetries
	}
}

func WithMaxChunkSize(maxChunkSize int) Option {
	return func(d *Downloader) {
		d.config.MaxChunkSize = maxChunkSize
	}
}

func WithConcurrency(concurrency int) Option {
	return func(d *Downloader) {
		d.config.Concurrency = concurrency
	}
}

func WithChunkSizeDeterminer(chunkSizeDeterminer ChunkSizeDeterminer) Option {
	return func(d *Downloader) {
		d.config.ChunkSizeDeterminer = chunkSizeDeterminer
	}
}

func WithPartDeterminer(partDeterminer PartDeterminer) Option {
	return func(d *Downloader) {
		d.config.PartDeterminer = partDeterminer
	}
}

// Download downloads a file from url and save it to fileOut
func (d *Downloader) Download(url string, fileOut string) error {
	return d.DownloadWithContext(context.Background(), url, fileOut)
}

// Like Download, but with context
//
// You can use context to cancel the download
func (d *Downloader) DownloadWithContext(ctx context.Context, url string, fileOut string) error {
	// TODO: Implement this function
	return nil
}
